### PR TITLE
Expose server build information

### PR DESCRIPTION
**Pull request by kashike**
Server startup & `/version`:
>  This server is running Paper version 1.20.6-DEV-feature/better-version@c1d8c2e (2024-05-15T21:19:32Z) (Implementing API version 1.20.6-R0.1-SNAPSHOT)
> This server is running Paper version 1.20.6-205-feature/better-version@c1d8c2e (2024-05-15T21:21:30Z) (Implementing API version 1.20.6-R0.1-SNAPSHOT)

---

`Bukkit.getVersion()`:
> 1.20.6-DEV-c1d8c2e (MC: 1.20.6)
> 1.20.6-205-c1d8c2e (MC: 1.20.6)

---

`DEV` will be replaced with a real build number (ex: `204`) when built using TeamCity.

### CODE REVIEW

File: patches/api/0480-Expose-server-build-information.patch [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
++     */
++    @NotNull String apiVersion();
++
++    /**
++     * Creates a string representation of the server build information.
++     *
++     * @param representation the type of representation
++     * @return a string
++     */
++    @NotNull String toString(final @NotNull StringRepresentation representation);
```
Comment (Machine-Maker) [inline patches/api/0480-Expose-server-build-information.patch]: I like `asString` better for this kind of method

File: patches/api/0480-Expose-server-build-information.patch [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+     public static String getName() {
+@@ -166,6 +160,8 @@ public final class Bukkit {
+      * Gets the version of game this server implements
+      *
+      * @return version of game
++     * @see io.papermc.paper.ServerBuildInfo#minecraftVersionId()
++     * @see io.papermc.paper.ServerBuildInfo#minecraftVersionName()
+      */
+     @NotNull
+     public static String getMinecraftVersion() {
```
Comment (Machine-Maker) [inline patches/api/0480-Expose-server-build-information.patch]: Probably add an `@see` to Bukkit.getVersion()

File: patches/api/0480-Expose-server-build-information.patch [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
++     * Gets the {@code ServerBuildInfo}.
++     *
++     * @return the {@code ServerBuildInfo}
++     */
++    static @NotNull ServerBuildInfo get() {
++        //<editor-fold defaultstate="collapsed" desc="Holder">
++        final class Holder {
++            static final ServerBuildInfo INSTANCE = Services.service(ServerBuildInfo.class).orElseThrow();
++        }
++        //</editor-fold>
```
Comment (Machine-Maker) [inline patches/api/0480-Expose-server-build-information.patch]: Did we confirm that simply classloading ServerBuildInfo doesn't classload the method local class and throw an exception if no service exists?

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
Looks mostly good. And when its done we have to merge the several patches that all touch version stuff and probably put this at the beginning of the patches.

File: patches/api/0480-Expose-server-build-information.patch [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
++     * The brand id for Paper.
++     */
++    Key BRAND_PAPER_ID = Key.key("papermc", "paper");
++
++    /**
++     * Gets the {@code ServerBuildInfo}.
++     *
++     * @return the {@code ServerBuildInfo}
++     */
++    static @NotNull ServerBuildInfo get() {
```
Comment (Machine-Maker) [inline patches/api/0480-Expose-server-build-information.patch]: Are we sure about `get`? The other similar interface we added, `RegistryAccess`'s static getter method is `registryAccess()`. Am not the biggest fan of `get` due to static imports.

**Review by Machine-Maker — APPROVED** [review]

**Comment by kashike** [comment]
FYI: We're going to be squashing a bunch of previous patches dealing with versioning into one mega-patch.

**Pull request merged**